crates/common/src/opcode.rs

```diff
@@ -80,6 +87,16 @@ impl TryFrom<M31> for Opcode {
     }
 }
 
+impl TryFrom<QM31> for Opcode {
+    type Error = InstructionError;
+
+    fn try_from(value: QM31) -> Result<Self, Self::Error> {
+        let opcode_u32 = value.to_m31_array()[0].0;
```

Comment:
you need to assert that all other values are 0. You can just use the fact that `impl TryInto<M31> for QM31 {` exists, so `value.try_into()?.into()` should be ok